File: src/effs/socket_close.js
// Socket
// ======
//! use ./sys.js

function socket_close(socket) {
  const sys = sys_get();
  const tcp = sys.read(socket, "tcp");
  const udp = sys.read(socket, "udp");
  if (tcp !== null || udp !== null) {
    sys.s.close(sys.kill(socket));
  }
  return { $: "Unit" };
}

File: src/effs/socket_close.c
// Socket
// ======
//! use ./sys.c

void socket_close(IoHand socket) {
  int tcp = io_sys_read(socket, IO_TCPS);
  int udp = io_sys_read(socket, IO_UDPS);
  if (tcp >= 0 || udp >= 0) {
    close(io_sys_kill(socket));
  }
}

Term socket_close_run(Env e, Term* f) {
  socket_close(io_hand_c(e, f[0]));
  return term_pak(CID_UNIT, 0);
}

static void __attribute__((constructor)) socket_close_use(void) {
  io_eff(FID_SOCKET_CLOSE, CID_SOCKET_CLOSE, socket_close_run);
}
